ros/src/twist_controller/twist_controller_mod.py

```diff
@@ -82,12 +86,16 @@ def control(self, target_linear_velocity, target_angular_velocity, current_linea
             brake_command = self.braking_pid_controller.step(-velocity_error, timestep) if velocity_error < (-1 * limit_constant *
                                                                                                              self.braking_to_throttle_threshold_ratio * current_linear_velocity) or (velocity_error < 0 and current_linear_velocity < 2.5) else 0
 
+        # apply low pass filter on throttle and brake commands to reduce excessive jerk
+        filtered_throttle = self.throttle_lpf.filt(throttle_command)
+        filtered_brake = self.brake_lpf.filt(brake_command)
+
         rospy.logdebug('Current linear velocity %.2f, target linear velocity %.2f, throttle_command %.2f, brake_command %.2f',
-                       current_linear_velocity, target_linear_velocity, throttle_command, brake_command)
+                       current_linear_velocity, target_linear_velocity, filtered_throttle, filtered_brake)
 
         # Return throttle, brake, steer
-        return (throttle_command,
-                brake_command,
+        return (filtered_throttle,
+                filtered_brake if filtered_throttle > 0 else 0,
```

Comment:
I think this is supposed to be `if filtered_throttle == 0` but actually the filtering makes it so it almost never completely reaches 0.  We can force the filtered values to hard 0 if they get below a small threshold, but I'm still thinking about the best way to handle the trade offs between throttle and brake.  I'm going to try some more testing tomorrow.

Reply:
Yes, it is incorrect - I already found that issue last night, and fixed it locally, but other issues cropped up which took my attention away from checking in the fix immediately :)

It seems that my PID control tuning that I applied a week or so back is no longer valid at lower speeds (5m/s), so I'm re-tuning the controllers now. Hopefully I'll be able to get those retuned again and won't need to return to them after that. Stay tuned... :P

Reply:
I also noticed some throttle/brake hunting at low target speeds and have a few ideas that might be contributing.  I haven't confirmed with the data but you might want to consider these before retuning the PID gains:

1. Since we are switching between activating the throttle/brake PID output, the integral terms need to be reset and kept from winding up while each controller's output is not active.  This means when the throttle PID is not active, we can keep it reset so that when it reactivates again the integral term will start from zero, and same for the brake PID controller.

2. The hysteresis thresholds for switching between throttle and brake control are percentage based, so when the target speed is very small, the hysteresis becomes very small and it is easy to get activation hunting again.  Maybe the thresholds should be based on absolute m/s values instead to keep big enough gaps to prevent the hunting.  

3.  The limit_constant threshold size switches based on velocity error being positive or negative, but actually it should probably switch based on the control mode itself, for example when it crosses the threshold from accel control ON to OFF, the threshold for activating accel control should then change to prevent it easily changing the mode back ON.  I think either the mode can be inferred by if the last output, like if throttle command is zero (accel mode) or not (brake mode), or we could separate the control mode judgment into its own conditional block to make it more clear, and then use the control mode to switch between the throttle or brake output.

Reply:
> Since we are switching between activating the throttle/brake PID output, the integral terms need to be reset and kept from winding up while each controller's output is not active. This means when the throttle PID is not active, we can keep it reset so that when it reactivates again the integral term will start from zero, and same for the brake PID controller.

This is the solution that I was looking for. I'd like to believe that I would have found this before you mentioned it, but you beat me to the punch.

This seems to have corrected all of the issues that I'm currently seeing. I'm going to add this to the PR and merge it into master; I'll keep the other two suggestions in my back pocket for future robustness!